File: ConsoleApplication/Tree/SubtreeCheck.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApplication1.Tree
{
    public class SubtreeCheck
    {
        private bool AreTreesSame(TreeNode t, TreeNode s)
        {
            if (t == null && s == null)
            {
                return true;
            }

            if (t == null || s == null)
            {
                return false;
            }

            return t.Data == s.Data && AreTreesSame(t.Left, s.Left) && AreTreesSame(t.Right, s.Right);
        }

        public bool IsSubtree(TreeNode t, TreeNode s)
        {
            if (t == null)
            {
                return false;
            }

            if (s == null)
            {
                return true;
            }

            if (AreTreesSame(t, s))
            {
                return true;
            }

            return IsSubtree(t.Left, s) || IsSubtree(t.Right, s);
        }



        bool AreTreesEqual2(TreeNode t, TreeNode s)
        {
            if (t == null && s == null)
            {
                return true;
            }
            
            if (t == null || s == null)
            {
                return false;
            }

            return t.Data == s.Data && AreTreesEqual2(t.Left, s.Left) && AreTreesEqual2(t.Right, s.Left);
        }

        public bool IsTreeASubtree(TreeNode t, TreeNode s)
        {
            if (t == null) return false;
            if (s == null) return true;

            if (AreTreesEqual2(s, t))
            {
                return true;
            }

            return IsTreeASubtree(t.Left, s) || IsTreeASubtree(t.Right, s);
        }
        
    }
}

File: Tasks/src/trees/SubtreeCheck.kt
package trees

class SubtreeCheck {

    fun check() {
        val t1 = Tree.getExampleTree1()
        val t2 = Tree.getExampleTree4()

        var result = false
        if (t2 == null) {
            result = true
        } else {
            result = subtreeCheck(t1, t2)
        }

        println(result)
    }

    private fun subtreeCheck(t1: TreeNode?, t2: TreeNode?): Boolean {

        if (t1 == null) {
            return false
        } else if (t1?.value == t2?.value && isSubtree(t1, t2)) {
            return true
        }

        return subtreeCheck(t1?.left, t2) || subtreeCheck(t1?.right, t2)
    }

    private fun isSubtree(t1: TreeNode?, t2: TreeNode?): Boolean {
        if (t1 == null && t2 == null) {
            return true
        } else if (t1 == null || t2 == null) {
            return false
        } else if (t1.value == t2.value) {
            return  isSubtree(t1.left, t2.left) && isSubtree(t1.right, t2.right)
        }

        return false
    }
}